dataset/shape.cpp

```diff
@@ -141,4 +141,69 @@ Dataset resize(const DatasetConstView &d, const Dim dim,
   return result;
 }
 
+namespace {
+
+bool all_dims_unchanged(const Dimensions &item_dims,
+                        const Dimensions &new_dims) {
+  for (const auto dim : item_dims.labels()) {
+    if (!new_dims.contains(dim))
+      return false;
+    if (new_dims[dim] != item_dims[dim])
+      return false;
+  }
+  return true;
+}
```

Comment:
Is this conceptually correct? In a `reshape` operation we map new dimensions onto an underlying buffer of data. There is no connection between old and new dimension labels. For example:
- reshaping `(x:10,y:10)` onto `(y:10,x:10)` would return `true` here for an `x` coord with `item_dims = {x:10}`.
- below:
  ```cpp
      if (all_dims_unchanged(coord.dims(), dims))
        reshaped.coords().set(name, coord);
  ```
  Would keep the `x` coord depending on `x`.
- I think the correct action should instead rename dim `x` to dim `y`.
- I have to think some more about the coord name, but I think it would just be kept, i.e., output would have an `x` coord depending on `y`.

Reply:
But then in the example that you give, the only way you know that the `x` coordinate has to be renamed to `y` is by comparing old and new dims (and their order)? So there effectively is a connection between old and new dims.
The special case of `(x:10,y:10)` is more of a `transpose` operation than a `reshape`. That said, you should still be able to do it with `reshape` and it should give you the correct result.
And I agree that my implementation would fail on this case.
But to get te behaviour you describe, i think there is no way around comparing old and new dims.

Reply:
I am not suggesting renaming coords.

Reply:
yes but when you rename a data variable, it's very simple. You check the two shape volumes, and you just set the variable dims to the new dims, as you say, without caring about the old dims.
When you have a `x` coord attached to your 2d variable, you need to realise that the `x` dim has been renamed to `y` by comparing the old and the new dims. I don't think you can do this without comparing old and new dims.

Reply:
I think we *have to* do it without, since comparing dims gives you wrong results. It switches the dim a coord is assigned to. Consider, with you implementation in the example above:

```python
reshaped['x', 5].coords['x'] # gives orignal x['x', 5] ... but linked to data row that was previously `y`!
```

What we have to do instead is consider only the *shapes*. We have to:
1. Infer from the input shapes an dims which output dims are relevant for a particular coord
2. Determine whether reshaping is possible. Essentially, whether the subset of data input dims that the original coord/mask/attr has maps cleanly onto a subset of the output dims, essentially whether there is a `M:N` mapping with integer `M` and `N`. Potentially need to consider broadcasting.
3. Drop if such a mapping is not possible.

Reply:
As I said above, my current implementation is indeed wrong in this case. I wasn't saying I wasn't going to change it because in this example it wrongly copies the `x` coordinate as is. As you say, it should be assigned to the `y` dim.
What i was saying was that the only way you can know which dims needs to be renamed to what, is by comparing `['x', 'y']` to `['y', 'x']` and notice that `x` has been renamed to `y`. I don't think you can achieve this by ignoring the old dims and just looking at the new dims.

And i think it's also not so clear what to do when you are reshaping `('x':10, 'y':10)` to
```Py
sc.reshape(a, dims=['z', 'y', 'tof'], shape=(2, 10, 5))
```
I guess you would have to drop all coords because if you keep `y` it will be associated to different data values?
But do you keep coord `y` if
```Py
sc.reshape(a, dims=['z', 'tof', 'y'], shape=(2, 5, 10))
```
?

Reply:
> What i was saying was that the only way you can know which dims needs to be renamed to what, is by comparing ['x', 'y'] to ['y', 'x'] and notice that x has been renamed to y. I don't think you can achieve this by ignoring the old dims and just looking at the new dims.

I disagree, but probably best explained in a call. But what I mean boils down to:
1. Look at the output of `sc.show(data)`.
2. Consider *only* the shapes and their *relative relations* (e.g., "vertical" coord). *Ignore* dim labels.
3. Map new dim labels onto this.